sql: Do not truncate Decimal values in-place

andy-kimball opened the pull request:
Currently, the CheckValueWidth method truncates the incoming Decimal value
if it doesn't fit in the prescribed column type. Currently, the truncation
modifies the value in-place. In the INSERT/UPDATE code paths, this results
in the original AST DDecimal Datum node being modified. While this is not
currently a bug, since the AST is only used on a single thread, it will
be a problem for the query cache, which can read Datum values on multiple
threads. In addition, it breaks interning assumptions in the optimizer,
because the same Datum object cannot be reused when the query contains
duplicate values, because its value can unexpectedly change.

The fix is to make a copy of the Decimal when it doesn't fit the target
column type. This commit renames CheckValueWidth to LimitValueWidth and
adds an output value that will be a copy of the value in the truncation
case.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32369)
<!-- Reviewable:end -->

knz reviewed (approved):
Functionally LGTM, but do you not think it would be worth to avoid the allocation of the `outVal` on the heap in the (common) case where no truncation is needed?

andy-kimball reviewed (commented):
I should be falling through and returning `inVal` as the `outVal` when no truncation is needed. Where do you see an allocation occurring in that code path?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-:-LRKdsw7AgIOuMYIqA-i:b22280o)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/eval.go, line 3074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-LRLdR4HCw5yNH-ZFESn:-LRLdR4IBJ0VMFRBl8CH:b-ehfesv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f57bf162e8b8102ea37c0195dd596c0aa96d817/pkg/sql/sem/tree/eval.go#L3074)):*
> ```Go
> 			var limited DDecimal
> 			err = LimitDecimalWidth(&dd.Decimal, &limited.Decimal, typ.Prec, typ.Scale)
> 			return &limited, err
> ```

This could avoid an alloc in the happy case where no truncation is needed.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/eval.go, line 3074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-LRLdR4HCw5yNH-ZFESn:-LRN274k6nE4IdvDRFEm:b-qdgp5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f57bf162e8b8102ea37c0195dd596c0aa96d817/pkg/sql/sem/tree/eval.go#L3074)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This could avoid an alloc in the happy case where no truncation is needed.
</blockquote></details>

But something does need to escape, because a conversion has occurred. In the previous code, `dd` escaped. Now `limited` escapes instead. The change shouldn't cause any extra object to escape; it's just changing which object escapes is all.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/eval.go, line 3074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-LRLdR4HCw5yNH-ZFESn:-LRN6vHv85YvHVUgnzcF:b-eo3udi) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f57bf162e8b8102ea37c0195dd596c0aa96d817/pkg/sql/sem/tree/eval.go#L3074)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But something does need to escape, because a conversion has occurred. In the previous code, `dd` escaped. Now `limited` escapes instead. The change shouldn't cause any extra object to escape; it's just changing which object escapes is all.
</blockquote></details>

If my reading is right, in the current code, both `dd` and `limited` escape. `dd` escapes because `val` takes an address "inside" `dd` and is subsequently used as receiver for some method calls.
If there was a way to check the width in the result in `dd` this would restrict the allocations to just 1.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/eval.go, line 3074 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-LRLdR4HCw5yNH-ZFESn:-LRNDuPP1Gn-qs87cBAn:b-eaywr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f57bf162e8b8102ea37c0195dd596c0aa96d817/pkg/sql/sem/tree/eval.go#L3074)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

If my reading is right, in the current code, both `dd` and `limited` escape. `dd` escapes because `val` takes an address "inside" `dd` and is subsequently used as receiver for some method calls.
If there was a way to check the width in the result in `dd` this would restrict the allocations to just 1.
</blockquote></details>

Hm, I'd used `goescape` and didn't see that, but I just tried again and it does escape after all. I must have missed it first time. I fixed by changing the code to use `tmpDec` in the `EvalContext`. Good catch.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-:-LRNOJ0291_Ty7FtqakF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andy-kimball committed:
sql: Do not truncate Decimal values in-place

Currently, the CheckValueWidth method truncates the incoming Decimal value
if it doesn't fit in the prescribed column type. Currently, the truncation
modifies the value in-place. In the INSERT/UPDATE code paths, this results
in the original AST DDecimal Datum node being modified. While this is not
currently a bug, since the AST is only used on a single thread, it will
be a problem for the query cache, which can read Datum values on multiple
threads. In addition, it breaks interning assumptions in the optimizer,
because the same Datum object cannot be reused when the query contains
duplicate values, because its value can unexpectedly change.

The fix is to make a copy of the Decimal when it doesn't fit the target
column type. This commit renames CheckValueWidth to LimitValueWidth and
adds an output value that will be a copy of the value in the truncation
case.

Release note: None

andy-kimball reviewed (commented):
My latest changes caused a test failure, due to attempted reuse of the temp decimal value without first deep copying it. I've decided to revert the temp decimal change, as well as the original `LimitValueWidth` change. The design of the `apd.Decimal` class allows it to be modified in-place. Instead, I changed the column width checking code to always create a copy of the decimal if it might be limited. I also fixed a small bug in the decimal conversion logic, which was not performing a deep copy of the input decimal value.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-:-LRSAUEkEG-m15Tf2BLm:bug58nf)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32369#-:-LRSFa4c5MbBA4gVDJcF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+
